src/opaque-types-type-alias-impl-trait-inference.md

```diff
@@ -0,0 +1,193 @@
+# Inference of opaque types (type alias `impl Trait`)
+
+This page describes how the compiler infers the hidden type for an opaque type. This kind of type inference is particularly complex because, unlike other kinds of type inference, it works across functions and function bodies.
+
+## Running example
+
+To help explain how it works, let's start with a simple example.
+
+###
+
+```rust
+mod m {
+    pub type Seq<T> = impl IntoIterator<Item = T>;
+    
+    pub fn produce_singleton<T>(t: T) -> Seq<T> { 
+        vec![t]
+    }
+
+    pub fn produce_doubleton<T>(t: T, u: T) -> Seq<T> { 
+        vec![t, u]
+    }
+}
+
+fn is_send<T: Send>(_: &T) {}
+
+pub fn main() {
+    let elems = m::produce_singleton(22);
+    
+    is_send(&elems);
+
+    for elem in elems {
+        println!("elem = {:?}", elem);
+    }
+}
+```
+
+* In this example, the opaque type is `Seq<T>`:
+    * Its defining scope is the module `m`.
+    * Its hidden type is  `Vec<T>`, which is inferred from:
+        * `m::produce_singleton`
+        * `m::produce_doubleton`
+* In the `main` function, the opaque type is out of scope:
+    * When `main` calls `m::produce_singleton`, it gets back a reference to the opaque type `Seq<i32>`.
+    * The `is_send` call checks that `Seq<i32>: Send`. `Send` is not listed amongst the bounds of the impl trait, but because of auto-trait leakage, we are able to infer that it holds.
+    * The for loop desugaring requires that `Seq<T>: IntoIterator`, which is provable from the bounds declared on `Seq<T>`.
```

Comment:
I'd suggest making each of the top-level bullet points here its own paragraph, with a sentence each for the level below.